### PR TITLE
New UI library, some bug fixes and features

### PR DESCRIPTION
**Pull request by jherico**
This PR moves the base UI code out of render-utils (where it was originally create because it needed an offscreen GL canvas, which was supplied by some new classes created in render-utils) into it's own library, where we can place common logic and reusable controls that aren't specific to the interface functionality (like message boxes).  

There is also some refactoring of the names of our QML files and their organization to try to break things up by category rather than having a single giant folder of QML files doing various jobs.  Many of the old files with names like 'CustomText' were previously just copies of work I had done on Shadertoy VR without much thought to maintainability outside of myself.

Also fixes the enter key for accessing the address bar.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: dab8761c714495c14c2ed3192412bd89d0c2a5bc available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: eb02bdc67f7071519cae94dbaf8c52c12dc995cc available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fdf2f8395d0949b5cfaf4f55eb21287f91a281c8 available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9fbd5d79e4e2bf34c6f7fbb5a665d318322c6bae available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

File: libraries/ui/src/HifiMenu.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
+#include "HifiMenu.h"
```
Comment (ZappoMan) [inline libraries/ui/src/HifiMenu.cpp]: missing file header

File: libraries/ui/src/HifiMenu.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
+//
+//  MenuConstants.h
```
Comment (ZappoMan) [inline libraries/ui/src/HifiMenu.h]: wrong file name

File: libraries/ui/src/HifiMenu.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+//
+//  Created by Bradley Austin Davis on 2015/04/21
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#pragma once
+#ifndef hifi_MenuContants_h
```
Comment (ZappoMan) [inline libraries/ui/src/HifiMenu.h]: file name

File: libraries/ui/src/HifiMenu.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    QHash<QString, std::function<void(bool)>> _toggleActions;
+    QObject* findMenuObject(const QString& name);
+    const QObject* findMenuObject(const QString& name) const;
+    QObject* _rootMenu{ nullptr };
+    QSignalMapper _triggerMapper;
+    QSignalMapper _toggleMapper;
+};
+
+#endif // hifi_MenuConstants_h
+
```
Comment (ZappoMan) [inline libraries/ui/src/HifiMenu.h]: there are only so many blank lines to go around.... these guys seem to be ganging up on us

File: libraries/ui/src/MessageDialog.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
+//
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.cpp]: extra line?

File: libraries/ui/src/MessageDialog.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+#include "MessageDialog.h"
+
+HIFI_QML_DEF(MessageDialog)
+
+MessageDialog::MessageDialog(QQuickItem *parent) : OffscreenQmlDialog(parent) {
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.cpp]: coding standard: type squish

File: libraries/ui/src/MessageDialog.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+MessageDialog::Icon MessageDialog::icon() const {
+    return _icon;
+}
+
+void MessageDialog::setVisible(bool v) {
+    OffscreenQmlDialog::setVisible(v);
+}
+
+void MessageDialog::setText(const QString &arg) {
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.cpp]: coding standard: type squish

File: libraries/ui/src/MessageDialog.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    if (arg != _text) {
+        _text = arg;
+        emit textChanged();
+    }
+}
+
+void MessageDialog::setInformativeText(const QString &arg) {
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.cpp]: coding standard: type squish

File: libraries/ui/src/MessageDialog.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    if (arg != _informativeText) {
+        _informativeText = arg;
+        emit informativeTextChanged();
+    }
+}
+
+void MessageDialog::setDetailedText(const QString &arg) {
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.cpp]: coding standard: type squish

File: libraries/ui/src/MessageDialog.h [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    QString _detailedText;
+    Icon _icon{ Information };
+    StandardButtons _buttons;
+    StandardButton _clickedButton{ NoButton };
+    OffscreenUi::ButtonCallback _resultCallback;
+};
+
+#endif // hifi_MessageDialog_h
+
+
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.h]: extra lines

File: libraries/ui/src/MessageDialog.h [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    virtual ~MessageDialog();
+
+    QString text() const;
+    QString informativeText() const;
+    QString detailedText() const;
+    Icon icon() const;
+
+public slots:
+    virtual void setVisible(bool v);
+    void setText(const QString &arg);
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.h]: coding standard: type squish

File: libraries/ui/src/MessageDialog.h [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    void setInformativeText(const QString &arg);
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.h]: coding standard: type squish

File: libraries/ui/src/MessageDialog.h [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    void setDetailedText(const QString &arg);
```
Comment (ZappoMan) [inline libraries/ui/src/MessageDialog.h]: coding standard: type squish

File: libraries/ui/src/OffscreenQmlDialog.cpp [hunk 14]
```diff
@@ -13,6 +13,30 @@
...
+
 void OffscreenQmlDialog::hide() {
     static_cast<QQuickItem*>(parent())->setEnabled(false);
 }
+
+QString OffscreenQmlDialog::title() const {
+    return _title;
+}
+
+void OffscreenQmlDialog::setTitle(const QString &arg) {
```
Comment (ZappoMan) [inline libraries/ui/src/OffscreenQmlDialog.cpp]: coding standard: type squish

File: libraries/ui/src/OffscreenQmlDialog.h [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    Q_DECLARE_FLAGS(StandardButtons, StandardButton)
+
+protected:
+    void hide();
+    virtual void accept();
+    virtual void reject();
+
+public:    
+    QString title() const;
+    void setTitle(const QString &arg);
```
Comment (ZappoMan) [inline libraries/ui/src/OffscreenQmlDialog.h]: coding standard: type squish

File: libraries/ui/src/OffscreenUi.cpp [hunk 16]
```diff
@@ -117,37 +122,47 @@ void OffscreenUi::resize(const QSize& newSize) {
...
-QQmlContext* OffscreenUi::qmlContext() {
-    if (nullptr == _rootItem) {
-        return _qmlComponent->creationContext();
-    }
-    return QQmlEngine::contextForObject(_rootItem);
+QQuickItem* OffscreenUi::getRootItem() {
+    return _rootItem;
 }
 
+//QQmlContext* OffscreenUi::qmlContext() {
```
Comment (ZappoMan) [inline libraries/ui/src/OffscreenUi.cpp]: dead code?

File: libraries/ui/src/OffscreenUi.cpp [hunk 17]
```diff
@@ -163,17 +178,18 @@ void OffscreenUi::requestRender() {
...
         QList<QQmlError> errorList = _qmlComponent->errors();
         foreach(const QQmlError &error, errorList) {
             qWarning() << error.url() << error.line() << error;
         }
-        return;
+        return nullptr;
     }
 
-    QObject* newObject = _qmlComponent->create();
+    QQmlContext * newContext = new QQmlContext(_qmlEngine, qApp);
```
Comment (ZappoMan) [inline libraries/ui/src/OffscreenUi.cpp]: coding standard: type squish

File: libraries/ui/src/OffscreenUi.cpp [hunk 18]
```diff
@@ -388,54 +410,68 @@ void OffscreenUi::setProxyWindow(QWindow* window) {
...
 }
 
 void OffscreenUi::messageBox(const QString& title, const QString& text,
+    ButtonCallback callback,
     QMessageBox::Icon icon,
-    QMessageBox::StandardButtons buttons,
-    ButtonCallback f) {
+    QMessageBox::StandardButtons buttons) {
+    MessageDialog * pDialog{ nullptr };
+    MessageDialog::show([&](QQmlContext*ctx, QObject*item) {
```
Comment (ZappoMan) [inline libraries/ui/src/OffscreenUi.cpp]: coding standard: spaces between type* and variable name

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0afe8c0e3268b0a92f09ebe382d635c7a08b05b2 available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2e61b5a933a502c8f32edddfc1f9710ee40b0194 available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3071be162ad935a82ec72a7212539866a4042c07 available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Comment by jherico** [comment]
should be ready for re-review.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a12bfd95f4167fa85c1a30f7e2c2dfbd96213688 available at: https://gist.github.com/60b408b79551d23c5739<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4695.dmg

**Pull request merged**